lib/features/build_server/cache/project_cache.dart

```diff
@@ -0,0 +1,235 @@
+      }
+
+      if (matches.isEmpty) {
+        required.add(id);
+      } else if (matches.isNotEmpty && matches.first.hash != checksum) {
+        _setIsCompiled(matches.first.id, false);
+        required.add(id);
+      } else {
+        inSync.add(id);
+      }
+    }
+
+    // Delete all files which are not required for update or already in sync.
+    // These files are not included in the latest project state and should therefore be deleted.
+    for (final entry in files.entries) {
+      if (!required.contains(entry.key) && !inSync.contains(entry.key)) {
+        final lock = _lock.files.singleWhere((file) => file.id == entry.key);
+
+        await File(lock.path).delete();
+        _lock.files.removeWhere((file) => file.id == entry.key);
+      }
+    }
+
+    return required;
+  }
+
+  /// Persists the passed [file].
+  Future<void> insert(SourceFile file) async {
+    final path = joinAll([
+      _workingDirectory.path,
+      'src',
+      '${file.entry.toString()}',
+      ...file.path,
+      '${file.name}.${file.extension}'
+    ]);
+    final diskFile = File(path);
+
+    await diskFile.create(recursive: true);
+    await diskFile.writeAsBytes(await file.data, flush: true);
+
+    if (_lock.files.where((f) => f.id == file.id).isNotEmpty) {
+      final lockFile = _lock.files.singleWhere((f) => f.id == file.id);
+      lockFile.hash = await file.hash;
+    } else {
+      _lock.files.add(SourceFileLock(id: file.id, hash: await file.hash, path: path));
+    }
+  }
+
+  /// A list of all stored header files.
+  List<File> get headerFiles {
+    final headers = <File>[];
+
+    for (final file in _lock.files) {
+      final ext = extension(file.path);
+
+      if (FileExtension.headers.contains(ext)) {
+        headers.add(File(file.path));
+      }
+    }
+
+    return headers;
+  }
+
+  /// A list of all stored source files.
+  ///
+  /// Each source file is returned as an object which contains
+  /// a path to the source file itself and to the object file.
+  /// The compiler must use this path to create the corresponding object file.
+  /// After compilation the `link` method has to be called, to let
+  /// the cache know that the latest version of the file has been compiled.
+  /// ```dart
+  /// final refs = await cache.sourceFiles;
+  ///
+  /// for (final ref in refs) {
+  ///   await compile(sourcePath: ref.source, objectPath: ref.object);
+  ///   await ref.link();
+  /// }
+  /// ```
+  Future<List<ImplementationObjectRef>> get sourceFiles async {
+    final refs = <ImplementationObjectRef>[];
+
+    for (final file in _lock.files) {
+      if (_isCompiled(file.id)) {
+        continue;
+      }
+
+      final ext = extension(file.path);
+
+      if (FileExtension.sources.contains(ext)) {
+        final sourceFile = File(file.path);
+        final objectFile = File(_objectPath(file.id));
+
+        await objectFile.create(recursive: true);
```

Comment:
More or less: yes

Reply:
This could be a little confusing tbh, empty (non working object-files). You could:
a) let the compiler create the directory (test if that works ^^)
b) just create the dir and not the file
c) assume noone will ever look :P

Reply:
I choose c :D The empty file 'problem' is not really a problem because I don't depend on the files to know that I've got the object file. That's managed in state and that's why we have `ImplementationObjectRef`